stp/stp.zkasm

```diff
@@ -0,0 +1,163 @@
+  gas_extra, stipend = call_gas_extra(INST, VALUE, WARM, EXISTS)
+  ;; determine upfront cost for call
+  c, GAS_UPFRONT = GAS_MXP + gas_extra
+  ;; sanity check precondition
+  if c == 1 goto failed_pre
+  ;; check for out-of-gas (oogx)
+  L_gas_diff, OOGX = gas_out_of_pocket(GAS_ACTUAL,GAS_UPFRONT)
+  ;; calculate gas_minc
+  GAS_OOP = min256_64(GAS,L_gas_diff)
+  ;; handle oogx case
+  if OOGX == 1 goto call_oogx
+  GAS_STIPEND = stipend
+  return
+call_oogx:
+  GAS_STIPEND=0
+  return
+  ;; ===========================================================================
+  ;; CREATE / CREATE2
+  ;; ===========================================================================
+create:
+  ;; determine upfront cost for create
+  c, GAS_UPFRONT = GAS_MXP + 32000
+  ;; sanity check precondition
+  if c == 1 goto failed_pre
+  ;; calculate out-of-pocket cost
+  GAS_OOP, OOGX = gas_out_of_pocket(GAS_ACTUAL,GAS_UPFRONT)
+  GAS_STIPEND = 0
+  return
+failed_pre:
+  ;; precondition failure
+  fail
+}
+
+;; Determines extra gas costs for a call instruction (such as for
+;; transfering value) and computes the gas stipend for the caller
+;; context.  The logic of this function is essentially determined by
+;; the yellow paper.
+fn call_gas_extra(inst=0xf1 u8, value u256, warm u1, exists u1) -> (gas_extra=2600 u64, stipend u64)
+;; PRE: inst in {0xf1,0xf2,0xf4,0xf5]
+{
+  var gas_transfer u16
+  var gas_new_account u16
+  var gas_access u16
+  ;;
+  if inst == 0xf1 goto call_only
+  if inst == 0xf2 goto call_other
+  if inst == 0xf4 goto call_no_transfer
+  if inst == 0xfa goto call_no_transfer
+  fail
+call_only:
+  if value == 0 goto call_no_transfer
+  if exists == 1 goto call_other
+  stipend = 2300 ;; G_callstipend
+  gas_transfer = 9000 ;; G_callvalue
+  gas_new_account = 25000 ;; G_newaccount
+  goto call
+call_other:
+  if value == 0 goto call_no_transfer
+  stipend = 2300 ;; G_callstipend
+  gas_transfer = 9000 ;; G_callvalue
+  gas_new_account = 0
+  goto call
+call_no_transfer:
+  stipend = 0
+  gas_transfer = 0
+  gas_new_account = 0
+call:
+  ;; check for warm access
+  if warm == 0 goto call_cold
+  gas_access = 100 ;; G_warmaccess
+  goto call_cont
+call_cold:
+  gas_access = 2600 ;; G_coldaccountaccess
+call_cont:
+  gas_extra = gas_access + gas_transfer + gas_new_account
+  return
+}
+
+;; Determine either: (a) the gas taken immediately when a call /
+;; create executes; or (b) that an out-of-gas exception occurred.
+;; Here, gas_actual is the actual amount of gas available prior to the
+;; call, whilst gas_upfront is the computed upfront cost in gas
+;; required for the call to proceed.
+fn gas_out_of_pocket(gas_actual u64, gas_upfront u64) -> (oop u64, oogx u1) {
+  var diff u64
+  var one_64th u58
+  var tmp u6
+  var z u1
+  ;; check for out-of-gas (oogx)
+  oogx, diff = gas_actual - gas_upfront
+  ;; split cases for oogx
+  if oogx == 1 goto exit_oogx
+  ;; calculate 1/64 of diff
+  one_64th, tmp = diff
```

Comment:
What's going on here ? Should there be a division by 64 on the right hand side ?

Reply:
Oh, I guess this is implicitly splitting stuff like the _harder case_ below suggests
```
   ;; harder case: split gas
   tmp2, res = gas
```
So I take it the type annotations for `one_64th`, `tmp` determine the intended splitting ?